### ISSUE TITLE
Need support for non-negative weight.

### ISSUE THREAD

**Issue by mega-optimus** [pytorch/pytorch#97452]
### 🚀 The feature, motivation and pitch

Sometimes we need to have weights that are non-negative, and simply using `F.relu(w)` doesn't work.

Because, during training `w` could become negative, and in that situation `F.relu(w)` has 0-gradient, and from then on `w` will no longer be updated and gets stuck there forever.

Current solution is to clamp `w` by 0 after gradient update in the training script, which is outside the module containing `w` and thus breaks encapsulation.

It would be nice to have first-class support for non-negative weights, marked by a `non-neg` attribute of `w`, and the system automatically clamps `w` by 0 after gradient update.

**Comment by vadimkantorov** [comment]
I wonder also if https://github.com/pytorch/pytorch/issues/7313 can be used for this...

**Comment by mega-optimus** [comment]
> I wonder also if #7313 can be used for this...

Thanks! Yes it can be related if implemented. I'm thinking of something simpler for bounded weights, see modified post above.

**Comment by vadimkantorov** [comment]
I also tagged @lezcano to comment on suitability of parametrizations for efficiently implementing this. I think, its API is a good match. Implementing some sort of quasi-relu you're proposing might be a separate feature request...

**Comment by ngimel** [comment]
relu has a well-defined differentiability behavior. If you need special non-standard definition of gradient, you should use custom autograd function or gradient hooks. 
Reparametrization functionality for parameters also exists, so I'm closing this issue, as it doesn't look like this feature should live in core.

**Comment by lezcano** [comment]
@ngimel is right here. Also, as a pointer, you may want to use LeakyReLU or play around with other activation functions that do not zero-out the inputs. FWIW, from a mathematical perspective, having incorrect gradients is not a good idea.

**Comment by mega-optimus** [comment]
understood, thanks!
